stdlib/public/core/FloatingPointTypes.swift.gyb

```diff
@@ -1810,6 +1817,89 @@ extension ${Self} : Strideable {
   }
 }
 
+//===----------------------------------------------------------------------===//
+// AnyHashable
+//===----------------------------------------------------------------------===//
+
+internal struct _${Self}AnyHashableBox: _AnyHashableBox {
```

Comment:
Since every float can be represented as a double, this seems like overkill. Just having a _DoubleAnyHashableBox would be fine, right?

Reply:
I'm still vacillating on whether AnyHashable should preserve the original type for numeric values. This version preserves them by using separate box types. Double doesn't have a generic initializer that takes a BinaryFloatingPoint value, so making the box generic would lead to ugly switches on the concrete type. 

I *think* it would be nice to ensure that for all stdlib numeric types, `(foo as AnyHashable).base` always returns the exact value we started with, of the original type. An alternative that is more consistent with SE-0170 is to allow `(42.0 as Double as AnyHashable).base` to return an integer, but I think I'd find that surprising.

I think that SE-0170 implies that we should allow downcasting a numeric AnyHashable to any standard integer/float type that can exactly represent the boxed value. I'm still working out how to best do that without going through NSNumber, but meanwhile, I want to make sure we can always downcast AnyHashable back to the original type. (Even if the Objective-C runtime isn't available, or if Foundation isn't loaded.)

```swift
let value: Num = 42 // where Num is any standard integer or floating point type
let boxed = value as AnyHashable
precondition(boxed.base is Num)
precondition(boxed as? Num != nil)
precondition(boxed as? Num == value)
```

Also, Float80 is a special case. It isn't bridged to NSNumber, which makes it the only standard numeric type that's unaffected by SE-0170. The current version of this PR rounds Float80 values to Double for AnyHashable comparisons, but that really should only be done if a Double can exactly represent the value.